pandas/core/internals/blocks.py

```diff
@@ -883,6 +883,14 @@ def setitem(self, indexer, value):
         ):
             values[indexer] = value
             try:
+                # GH25495 may need to convert to categorical block
```

Comment:
don't put complicated things inside a try/except it tends to hide errors

Reply:
this could be much earlier in the comparisons

Reply:
I've moved it out of the try/except. 

However, the intended behaviour is that we only convert to categorical if it is a full assignment, and only change the values in a partial assignment. Hence the block creation needs to be done after we've checked the shape of the new values matches existing values.